<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

The goal is to give a prediction of in which care unit will a patient be in in a certain number of days (or if they will be discharged). This way, hospital workload per care unit/department can be predicted. 

For this, I will have a table of patients with multiple features including medical data (diagnosis, medications...), previous stays and transfers, some demographic data, duration of stay... This can be done using the admissions table and joining information from the diagnosis table (specific per stay) 

At admission, which department will the patient end up in?
Will the patient visit this deparment in their stay?

Make more questions during the weekend.

Do the questions being clear about which variables you will use

And in two weeks have more or less the table ready. 

Before continuing, extract a smaller table that only contains relevant columns for finding atc1s. Find categories and fill them in that table. Then update/rejoin to prescriptions table and find transfer numbers.

I want to predict department occupancy in a hospital and I have a dataset that includes different information from patients. Initially at least, I am simplifying that by predicting instead the department that the patients are going to go to next. 

After a preliminary EDA I saw that I could not use the diagnosis data because it was not dated (it is a table with diagnosis for the whole stay), so now I am working to merge my database with a huge prescriptions table that is dated (so I can relate to patients for different stays and hospital transfers). I am converting the drugs to drug categories and did some frequency encoding, but there was like 20% of missing data, so I was filling that until recently. 

In principle I am working with 14 drug categories, but I have loads of data so could increase that to subcategories and apply PCA. In terms of modeling, my plan is to start with logistic regression and then move on to SVM, random forests...

In [1]:
import numpy as np
import pandas as pd
import copy
import nachopy
import importlib

In [2]:
importlib.reload(nachopy);

In [36]:
prescriptions = pd.read_csv('../Capstone data/MIMIC IV/prescriptions.csv')

C:\Users\Nacho\AppData\Local\Temp\ipykernel_1612\1308515471.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  prescriptions2 = pd.read_csv('../Capstone data/MIMIC IV/prescriptions.csv')


In [44]:
pres_dupl = prescriptions2[['subject_id','hadm_id','starttime','stoptime','drug']].duplicated()

In [48]:
prescriptions = prescriptions[~pres_dupl]

In [4]:
medrecon = pd.read_csv('../Capstone data/MIMIC IV/ed/medrecon.csv')

In [5]:
prescriptions['ndc'] = prescriptions.ndc.astype('Int64')
medrecon['ndc'] = medrecon.ndc.astype('Int64')

In [6]:
display(prescriptions.head())
display(prescriptions.info(show_counts = True))

subject_id   hadm_id  pharmacy_id       poe_id  poe_seq  \
0    10000032  22595853     11700683  10000032-34     34.0   
1    10000032  22595853     14779570  10000032-22     22.0   
2    10000032  22595853     19796602  10000032-50     50.0   
3    10000032  22595853     20256254  10000032-32     32.0   
4    10000032  22595853     28781051  10000032-27     27.0   

             starttime             stoptime drug_type  \
0  2180-05-07 01:00:00  2180-05-07 22:00:00      MAIN   
1  2180-05-07 00:00:00  2180-05-07 22:00:00      MAIN   
2  2180-05-08 08:00:00  2180-05-07 22:00:00      MAIN   
3  2180-05-07 01:00:00  2180-05-07 22:00:00      MAIN   
4  2180-05-07 00:00:00  2180-05-07 22:00:00      MAIN   

                          drug formulary_drug_cd     gsn          ndc  \
0                Acetaminophen           APAP500  004490    904198861   
1  Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN            0   
2                   Furosemide            FURO40  008209  51079007320   
3                  Raltegravir           RALT400  063231      6022761   
4                      Heparin            HEPA5I  006549  63323026201   

               prod_strength form_rx dose_val_rx dose_unit_rx form_val_disp  \
0               500mg Tablet     NaN         500           mg             1   
1              10 mL Syringe     NaN           3           mL           0.3   
2                40mg Tablet     NaN          40           mg             1   
3              400 mg Tablet     NaN         400           mg             1   
4  5000 Units / mL- 1mL Vial     NaN        5000         UNIT             1   

  form_unit_disp  doses_per_24_hrs  route  
0            TAB               NaN  PO/NG  
1            SYR               3.0     IV  
2            TAB               1.0  PO/NG  
3            TAB               2.0     PO  
4             mL               3.0     SC

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16219412 entries, 0 to 16219411
Data columns (total 20 columns):
 #   Column             Non-Null Count     Dtype  
---  ------             --------------     -----  
 0   subject_id         16219412 non-null  int64  
 1   hadm_id            16219412 non-null  int64  
 2   pharmacy_id        16219412 non-null  int64  
 3   poe_id             16091738 non-null  object 
 4   poe_seq            16091738 non-null  float64
 5   starttime          16201857 non-null  object 
 6   stoptime           16195756 non-null  object 
 7   drug_type          16219412 non-null  object 
 8   drug               16219411 non-null  object 
 9   formulary_drug_cd  16199404 non-null  object 
 10  gsn                14342210 non-null  object 
 11  ndc                16193185 non-null  Int64  
 12  prod_strength      16212752 non-null  object 
 13  form_rx            20731 non-null     object 
 14  dose_val_rx        16212798 non-null  object 
 15  dose_unit_rx 

None

In [7]:
display(medrecon.head())
display(medrecon.info(show_counts = True))

subject_id   stay_id            charttime  \
0    10000032  32952584  2180-07-22 17:26:00   
1    10000032  32952584  2180-07-22 17:26:00   
2    10000032  32952584  2180-07-22 17:26:00   
3    10000032  32952584  2180-07-22 17:26:00   
4    10000032  32952584  2180-07-22 17:26:00   

                                name    gsn          ndc  etc_rn  etccode  \
0                  albuterol sulfate  28090  21695042308       1   5970.0   
1                  calcium carbonate   1340  10135021101       1    733.0   
2       cholecalciferol (vitamin D3)  65241  37205024678       1    670.0   
3  emtricitabine-tenofovir [Truvada]  57883  35356007003       1   5849.0   
4          fluticasone [Flovent HFA]  21251  49999061401       1    371.0   

                                      etcdescription  
0  Asthma/COPD Therapy - Beta 2-Adrenergic Agents...  
1    Minerals and Electrolytes - Calcium Replacement  
2                           Vitamins - D Derivatives  
3  Antiretroviral - Nucleoside and Nucleotide Ana...  
4  Asthma Therapy - Inhaled Corticosteroids (Gluc...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3143791 entries, 0 to 3143790
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   subject_id      3143791 non-null  int64  
 1   stay_id         3143791 non-null  int64  
 2   charttime       3143791 non-null  object 
 3   name            3143791 non-null  object 
 4   gsn             3143791 non-null  int64  
 5   ndc             3143791 non-null  Int64  
 6   etc_rn          3143791 non-null  int64  
 7   etccode         3131412 non-null  float64
 8   etcdescription  3131412 non-null  object 
dtypes: Int64(1), float64(1), int64(4), object(3)
memory usage: 218.9+ MB


None

Warning about ndc having mixed types. Instead of using isna to find about null values, we will just take as valid NDCs those that are larger that have at least 6 digits (which is the smallest NDC according to the FDA liink). We can also check whether we have other NDC numbers with fewer digits, as these could be fixable mistakes.

Prescriptions and medrecon are similar,thankfully, although instead of hadm_id we have stay_id. And no null values on our drugs column, yay! We lack the drug type column.

In [8]:
def digit_checker(df, d_min = 0, d_max = 6):
    for i in range(d_min,d_max):
        d_no = 10**i

        validndc_bool = df.ndc>=d_no
        print(f'% of NDCs >= 10^{i}:', round(validndc_bool/len(df)*100,2))
        print('Number of NDCs:', df.ndc[validndc_bool].nunique(),'\n')

In [9]:
# display(prescriptions.nunique())
nachopy.digit_checker(prescriptions,d_max = 12)

% of NDCs >= 10^0: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^1: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^2: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^3: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^4: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^5: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^6: 87.56
Number of NDCs: 5744 

% of NDCs >= 10^7: 83.34
Number of NDCs: 5472 

% of NDCs >= 10^8: 77.31
Number of NDCs: 4727 

% of NDCs >= 10^9: 31.86
Number of NDCs: 2968 

% of NDCs >= 10^10: 31.3
Number of NDCs: 2952 

% of NDCs >= 10^11: 0.0
Number of NDCs: 0 



In [10]:
display(medrecon.nunique())
digit_checker(medrecon,d_max = 12)

subject_id        148919
stay_id           323582
charttime         581453
name               19165
gsn                 9262
ndc                 9398
etc_rn                 5
etccode             1208
etcdescription      1212
dtype: int64

% of NDCs >= 10^0: 0          0.0
1          0.0
2          0.0
3          0.0
4          0.0
          ... 
3143786    0.0
3143787    0.0
3143788    0.0
3143789    0.0
3143790    0.0
Name: ndc, Length: 3143791, dtype: Float64
Number of NDCs: 9397 

% of NDCs >= 10^1: 0          0.0
1          0.0
2          0.0
3          0.0
4          0.0
          ... 
3143786    0.0
3143787    0.0
3143788    0.0
3143789    0.0
3143790    0.0
Name: ndc, Length: 3143791, dtype: Float64
Number of NDCs: 9397 

% of NDCs >= 10^2: 0          0.0
1          0.0
2          0.0
3          0.0
4          0.0
          ... 
3143786    0.0
3143787    0.0
3143788    0.0
3143789    0.0
3143790    0.0
Name: ndc, Length: 3143791, dtype: Float64
Number of NDCs: 9397 

% of NDCs >= 10^3: 0          0.0
1          0.0
2          0.0
3          0.0
4          0.0
          ... 
3143786    0.0
3143787    0.0
3143788    0.0
3143789    0.0
3143790    0.0
Name: ndc, Length: 3143791, dtype: Float64
Number of NDCs: 9397 



So all NDCs larger than 1 appear to be correct. There are 12.44% missing NDC, but we will not be filling these. We are interested on the ATR4 classification, so we will match the available NDCs to this class and then try to fill any empty values for that feature.

For the NDC matching we are using a function provided by tal that queries tal tal. The prescriptions table was used as input for this function and the output is uploaded and mapped into our table here.

In [11]:
prescriptions.drop(columns = ['prod_strength','form_rx','dose_val_rx','dose_unit_rx','form_val_disp',\
                              'form_unit_disp','doses_per_24_hrs','pharmacy_id','poe_id','poe_seq','stoptime'],inplace = True)


Dropping some of the columns as they (may) have no relevance as we are substituting the drug features by their categories. This will also allow us to combine our dfs.

In [12]:
print(sum(medrecon.duplicated()))
dupl_rows = medrecon[['stay_id','charttime','name']].duplicated()
print(sum(dupl_rows))

0
439227


Medications are duplicated to include the different etc codes for one drug. That is okay, we could create dummies and group the columns. However, we will not use the edc in principle. We are going to move to ATCs, which are more interpretable, general and well-recognised. More importantly, they have a very customisable granularity, as we can select the level. At level 0 we have only 14 features, which simplifies our model greatly. We will start with that and increase granularity if necessary.

In [13]:
medrecon.drop(columns =  ['etccode','etc_rn','etcdescription'], inplace = True)
medrecon = medrecon[~dupl_rows]
medrecon.rename(columns = {'charttime':'starttime','name':'drug'},inplace = True)

In [14]:
prescriptions

subject_id   hadm_id            starttime drug_type  \
0           10000032  22595853  2180-05-07 01:00:00      MAIN   
1           10000032  22595853  2180-05-07 00:00:00      MAIN   
2           10000032  22595853  2180-05-08 08:00:00      MAIN   
3           10000032  22595853  2180-05-07 01:00:00      MAIN   
4           10000032  22595853  2180-05-07 00:00:00      MAIN   
...              ...       ...                  ...       ...   
16219407    19999987  23865745  2145-11-03 00:00:00      MAIN   
16219408    19999987  23865745  2145-11-04 10:00:00      MAIN   
16219409    19999987  23865745  2145-11-10 10:00:00      MAIN   
16219410    19999987  23865745  2145-11-09 10:00:00      MAIN   
16219411    19999987  23865745  2145-11-09 10:00:00      MAIN   

                                 drug formulary_drug_cd     gsn          ndc  \
0                       Acetaminophen           APAP500  004490    904198861   
1         Sodium Chloride 0.9%  Flush         NACLFLUSH     NaN            0   
2                          Furosemide            FURO40  008209  51079007320   
3                         Raltegravir           RALT400  063231      6022761   
4                             Heparin            HEPA5I  006549  63323026201   
...                               ...               ...     ...          ...   
16219407      Influenza Virus Vaccine         INFL0.5LF  069637  33332001201   
16219408         Levothyroxine Sodium            LEVO50  006649     74455211   
16219409                   Duloxetine            DULO30  057892      2324033   
16219410               Venlafaxine XR         VENL150XR  046405  68084048601   
16219411                   Duloxetine            DULO30  057892      2324033   

          route  
0         PO/NG  
1            IV  
2         PO/NG  
3            PO  
4            SC  
...         ...  
16219407     IM  
16219408  PO/NG  
16219409     PO  
16219410     PO  
16219411     PO  

[16219412 rows x 9 columns]

In [15]:
nachopy.display_sbs(titles = medrecon.stay_id.describe(),prescriptions.hadm_id.describe(),\
                    ['Stay_id from medrecon','Hadm_id from prescriptions'])

Stay_id from medrecon 
 
 
   
 stay_id 
 
 
 
 
 count 
 2733573.000000 
 
 
 mean 
 35001782.253237 
 
 
 std 
 2885614.265219 
 
 
 min 
 30000012.000000 
 
 
 25% 
 32502391.000000 
 
 
 50% 
 35000450.000000 
 
 
 75% 
 37498150.000000 
 
 
 max 
 39999964.000000 
 
 
 

 
 Hadm_id from prescriptions 
 
 
   
 hadm_id 
 
 
 
 
 count 
 16219412.000000 
 
 
 mean 
 25004407.280042 
 
 
 std 
 2884617.408397 
 
 
 min 
 20000019.000000 
 
 
 25% 
 22514037.000000 
 
 
 50% 
 25008728.000000 
 
 
 75% 
 27498545.000000 
 
 
 max 
 29999928.000000

Lovely, comparable, no overlap. We will combine them, adding a new column to indicate rows where hadm_id is actually stay_id (for info not for modelling).

In [50]:
medrecon.drop(columns = 'stay_id_ind',inplace = True)

In [51]:
prescriptions = pd.concat([prescriptions,medrecon.rename(columns = {'stay_id':'hadm_id'})],axis = 0)
prescriptions.reset_index(inplace = True,drop = True)

In [52]:
print(sum(prescriptions.duplicated()))
dupl_rows = prescriptions[['subject_id','hadm_id','starttime','stoptime','drug']].duplicated()
print(sum(dupl_rows))

0
0


In [18]:
prescriptions['starttime'] = pd.to_datetime(prescriptions.starttime)

In [19]:
import datetime

In [54]:
len(prescriptions[prescriptions.subject_id == 12468016])

8161

In [34]:
nachopy.display_sbs([prescriptions.ndc[prescriptions.stay_id_ind == 1].describe(),\
                      prescriptions.ndc[prescriptions.stay_id_ind == 0].describe()],\
                      titles = ['NDCs from hospital','NDCs from ED'])

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

It seems like ED and the hospital have completely different suppliers, because there is little overlap of ndcs. Let's export the unique ndc list to query the ATCs.

In [37]:
np.char.zfill(prescriptions.ndc.unique().astype('str'),11)

array(['00904198861', '00000000000', '51079007320', ..., '10939016244',
       '11917010387', '43376031060'], dtype='<U11')

In [ ]:
pd.Series(name = 'NDC',data= np.char.zfill(prescriptions.ndc.unique().astype('str'),11)).to_csv('unique_ndcs.csv')

In [57]:
mapped_NDC = pd.read_csv('../Capstone data/ndc_map/output/atc4/ndc_map 2022_10_12 17_12 (atc4).csv',usecols = ['ndc','atc4','atc4_name'])

In [58]:
mapped_NDC.dropna(inplace = True)
mapped_NDC.ndc = mapped_NDC.ndc.astype('int64')

In [59]:
mapped_NDC.drop_duplicates(inplace = True)
len(mapped_NDC)

21710

In [60]:
mapped_NDC.head()

ndc   atc4                                  atc4_name
2  2050301  J01GB                      Other aminoglycosides
3  2050301  S01AA                                Antibiotics
4  2060440  J04AB                                Antibiotics
5  2100902  R05DA            Opium alkaloids and derivatives
8  2143301  A10BJ  Glucagon-like peptide-1 (GLP-1) analogues

In [61]:
mapped_NDC.atc4_name.nunique()

518

A lot of categories. Feasible, but we will first try taking just the first letter (lvl 1 of classification) of the atc_4 instead

In [62]:
atc1 = mapped_NDC[['ndc','atc4']].copy()
atc1['atc4'] = atc1.atc4.str[0]
atc1.rename(columns = {'atc4':'atc1'},inplace = True)
atc1.drop_duplicates(inplace = True)

In [63]:
atc1.head()
print(len(atc1))
print(atc1.nunique())
display(atc1.groupby('ndc').nunique().describe())

16937
ndc     10770
atc1       14
dtype: int64


atc1
count  10770.000000
mean       1.572609
std        1.181846
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max        8.000000

In [64]:
atc1 = pd.concat([atc1,pd.get_dummies(atc1.atc1,prefix = None)],axis = 1).drop(columns = 'atc1')
atc1 = atc1.groupby('ndc').any()*1
display(atc1.head())
print(atc1.shape)

A  B  C  D  G  H  J  L  M  N  P  R  S  V
ndc                                              
2050301  0  0  0  0  0  0  1  0  0  0  0  0  1  0
2060440  0  0  0  0  0  0  1  0  0  0  0  0  0  0
2100902  0  0  0  0  0  0  0  0  0  0  0  1  0  0
2143301  1  0  0  0  0  0  0  0  0  0  0  0  0  0
2143401  1  0  0  0  0  0  0  0  0  0  0  0  0  0

(10770, 14)


There is more than one atc1 per ndc code! We were going to use dummy variables in the future, so we will implement those now before the merge. That way we will have just one category per ndc. Also note that not all ndcs have been mapped, 490 out of 5744 are still unclassified. We will try to fill those after the merge.

In [65]:
prescriptions.reset_index(inplace = True)
prescriptions.rename(columns = {'index':'join_on_index'},inplace = True)
prescriptions.set_index('ndc',inplace = True)
cat_pres = prescriptions[['join_on_index','drug_type','drug','formulary_drug_cd','gsn']].join(atc1)
cat_pres.reset_index(inplace = True)

In [66]:
display(cat_pres.sample(10))
print('Missing categories %:',sum(cat_pres.A.isna())/len(cat_pres)*100)
cat = list('ABCDGHJLMNPRSV')
cat_pres[cat] = cat_pres[cat].fillna(-1).astype('int8')

print(cat_pres.info(show_counts = True))

ndc  join_on_index drug_type  \
9308855     904516561.0        2504329      MAIN   
14699936  51079096620.0        6085387      MAIN   
7168339     409665305.0        3543575      MAIN   
418017              0.0        3574543      BASE   
5783571     338070948.0       14351113      MAIN   
11185323  10544034130.0       17151427       NaN   
13747385  50419040201.0       15069592       NaN   
14662615  51079093020.0        3916219      MAIN   
5742074     338070341.0        6630053      MAIN   
16270188  63323026965.0       12987648      MAIN   

                                             drug formulary_drug_cd     gsn  \
9308855                                     Senna           SENN187  019964   
14699936                               Famotidine            PEPC20  011677   
7168339                        Potassium Chloride            KCL40I  001255   
418017                                         SW          KCLBASE2     NaN   
5783571                        Potassium Chloride           KCL10PM  015362   
11185323                               omeprazole               NaN   33530   
13747385  drospirenone-ethinyl estradiol [Ocella]               NaN   47787   
14662615                               Carvedilol           CARV625  028109   
5742074                        Potassium Chloride           KCL20PM  045309   
16270188                                 Propofol         PROP100IG  016796   

            A    B    C    D    G    H    J    L    M    N    P    R    S    V  
9308855   1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
14699936  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
7168339   1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
418017    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5783571   1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
11185323  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
13747385  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  
14662615  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
5742074   1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
16270188  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0

Missing categories %: 17.853161576245586
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17628928 entries, 0 to 17628927
Data columns (total 20 columns):
 #   Column             Non-Null Count     Dtype 
---  ------             --------------     ----- 
 0   ndc                17603412 non-null  object
 1   join_on_index      17628928 non-null  int64 
 2   drug_type          14924364 non-null  object
 3   drug               17628927 non-null  object
 4   formulary_drug_cd  14904840 non-null  object
 5   gsn                15998809 non-null  object
 6   A                  17628928 non-null  int8  
 7   B                  17628928 non-null  int8  
 8   C                  17628928 non-null  int8  
 9   D                  17628928 non-null  int8  
 10  G                  17628928 non-null  int8  
 11  H                  17628928 non-null  int8  
 12  J                  17628928 non-null  int8  
 13  L                  17628928 non-null  int8  
 14  M                  17628928 non-null  i

In [67]:
cat_pres['ndc'] = cat_pres.ndc.floordiv(100)
cat_pres.loc[cat_pres.ndc == 0,'ndc'] = pd.NA
cat_pres['ndc'] = cat_pres.ndc.astype('Int64')

Note that the new category dummy columns are floats after the join, not int8s. We have filled the NaNs as -1 and changed the data type of the columns. This is useful to decrease the size of the dataframe and it will also be useful to fill the missing values (which is 19.26% of the rows), as we will see below. We will fill these missing values by matching with other columns and we there are a few candidates column for this: drug, pharmacy_id, gsn and formulary_drug_cd. The goal will be to find the column with values shared by the largest amount of rows, so matching ATR1s is more likely. 

In [68]:
nndrug = cat_pres[cat_pres.drug.isna()].index.values[0]
display(cat_pres.loc[nndrug])

display(cat_pres[cat_pres.ndc == cat_pres.ndc.loc[nndrug]].head(1))

cat_pres.loc[nndrug,['drug', 'formulary_drug_cd']] = ('Simvastatin','SIMV40')

print('Value counts of drugs:\n',cat_pres.drug.value_counts())
cat_pres.drug = cat_pres.drug.str.lower()
drugs_vc = cat_pres.drug.value_counts()

clin_drug = cat_pres.drug.str.contains('clindamyci')
display(cat_pres[clin_drug].head(1))
cat_pres.drug[clin_drug] = 'clindamycin'

print('\n\nValue counts of drugs after making all letters lowercase:\n',drugs_vc)
print('\nDescription:\n',drugs_vc.describe())

print('\nNumber of drugs used only once:', sum(drugs_vc>1))

cat_pres['drug'] = cat_pres.drug.str.replace('[^a-zA-Z]','', regex = True)
drugs_vc = cat_pres.drug.value_counts()

print('\n\nValue counts of drugs after making all letters lowercase:\n',drugs_vc)
print('\nDescription:\n',drugs_vc.describe())

print('\nNumber of drugs used only once:', sum(drugs_vc>1))



ndc                  510790456
join_on_index         14845444
drug_type                 MAIN
drug                       NaN
formulary_drug_cd      SIMVIND
gsn                     016579
A                            0
B                            0
C                            1
D                            0
G                            0
H                            0
J                            0
L                            0
M                            0
N                            0
P                            0
R                            0
S                            0
V                            0
Name: 14378694, dtype: object

ndc  join_on_index drug_type         drug formulary_drug_cd  \
14358871  510790456            333      MAIN  Simvastatin            SIMV40   

             gsn  A  B  C  D  G  H  J  L  M  N  P  R  S  V  
14358871  016579  0  0  1  0  0  0  0  0  0  0  0  0  0  0

Value counts of drugs:
 Insulin                        557829
Sodium Chloride 0.9%  Flush    530628
0.9% Sodium Chloride           508792
Acetaminophen                  411170
Furosemide                     333397
                                ...  
*ramapril                           1
*adderral                           1
*ellipta                            1
*bisa-lax                           1
Lumigan                             1
Name: drug, Length: 25853, dtype: int64


ndc  join_on_index drug_type             drug formulary_drug_cd gsn  \
1742431  <NA>       15092293       NaN  *clindamycin 1%               NaN   0   

         A  B  C  D  G  H  J  L  M  N  P  R  S  V  
1742431 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1

C:\Users\Nacho\AppData\Local\Temp\ipykernel_1612\2913060166.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cat_pres.drug[clin_drug] = 'clindamycin'




Value counts of drugs after making all letters lowercase:
 insulin                                          557853
sodium chloride 0.9%  flush                      530628
0.9% sodium chloride                             508792
acetaminophen                                    441995
furosemide                                       359143
                                                  ...  
desog-e.estradiol/e.estradiol [pimtrea (28)]          1
desog-e.estradiol/e.estradiol [kimidess (28)]         1
*genov                                                1
*steralive                                            1
lumigan                                               1
Name: drug, Length: 22976, dtype: int64

Description:
 count     22976.000000
mean        767.275766
std       10722.416656
min           1.000000
25%           1.000000
50%           3.000000
75%          17.000000
max      557853.000000
Name: drug, dtype: float64

Number of drugs used only once: 13943


Value counts of 

In [69]:
empty_drug = cat_pres.drug ==u''
print(f'There are {sum(empty_drug)} empty drugs')
display(cat_pres[empty_drug].sample())

fdrugs_empty = cat_pres.formulary_drug_cd[empty_drug].drop_duplicates().values
empty_drug = cat_pres.formulary_drug_cd.isin(fdrugs_empty)

cat_pres.drug.where(~empty_drug,cat_pres.formulary_drug_cd,inplace = True)

print(f'{sum(drugs_vc == 1)/len(cat_pres)*100}% of values in drugs column appears only once')

There are 112 empty drugs


ndc  join_on_index drug_type drug formulary_drug_cd gsn  A  B  C  D  \
1743711  <NA>       15286317       NaN                    NaN   0 -1 -1 -1 -1   

         G  H  J  L  M  N  P  R  S  V  
1743711 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1

0.04262879739482741% of values in drugs column appears only once


There was just one missing value for the drug column, which was easily filled by checking elsewhere in the table. After this, we analysed the number of rows for each drug. Interestingly, the formulary_drug_code was different than the original for that row. 

The input of these fields does not seem consistent. For example "Clindamycin" is misspelt (fixed for this particular case), abbreviations may not be consistent and upper and lower case seem to be used with little care. This last issue could be fixed with str.lower(), though. 

This means too many groups so matching categories will be harder. Let us look at the other columns with drug identifiers.

In [70]:
print('gsn value counts:\n',cat_pres.gsn.value_counts())

fd_vc = cat_pres.formulary_drug_cd.value_counts()
print('\n\nformulary_drug_cd value counts:\n',fd_vc)
print('\nDescription:\n',fd_vc.describe())

fd_vc_5 = fd_vc[fd_vc<=5].value_counts()
print('\nNumber of formulary_drug_cd present on only 5 rows or less:\n',fd_vc_5)
print('\n% of rows with drugs used 5 times or less:', round(fd_vc_5.sum()/fd_vc.sum()*100,5))

gsn value counts:
 001210                          757416
006549                          320343
027413                          252990
008205                          247213
004489                          242690
                                 ...  
003746                               1
002485                               1
045006 064197                        1
8245                                 1
021723 021726 021725 021724          1
Name: gsn, Length: 15601, dtype: int64


formulary_drug_cd value counts:
 NACLFLUSH       530435
HEPA5I          320637
ACET325         243655
DOCU100         243283
SENN187         206732
                 ...  
VANC.000125L         1
VANCOD125L           1
VANC1.25L            1
LEVO2                1
SAGE547/PCB          1
Name: formulary_drug_cd, Length: 3789, dtype: int64

Description:
 count      3789.000000
mean       3933.713381
std       18224.888227
min           1.000000
25%          10.000000
50%          97.000000
75%        1118.0000

From these candidates, formulary_drug_cd, gsn, and drugs seems the best to find the right ATR1s, but they have many null values. Additionally, the gsn column has 1.8 M missing values and sometimes several codes in one field. We will have to work around this
The pharmacy_id has many unique values, so this variable is not specific to the same drug, just an inventory code specific per item.  

We want to decrease the number of unique values as this will increase our chances of filling missing ATR1 fields. All our columns have many null values, but we can recursively fill categories based on all the identifying columns.

In [71]:
len(cat_pres)

17628928

In [72]:
cat_pres[cat].sum(axis = 1).value_counts()

 1     9481544
-14    3147321
 2     2660944
 3     1582873
 4      350029
 5      297370
 6       87202
 7       21059
 8         586
dtype: int64

In [73]:
cat_pres.gsn.value_counts()

001210                          757416
006549                          320343
027413                          252990
008205                          247213
004489                          242690
                                 ...  
003746                               1
002485                               1
045006 064197                        1
8245                                 1
021723 021726 021725 021724          1
Name: gsn, Length: 15601, dtype: int64

In [74]:
nachopy.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 17.853161576245586
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 16.9409058505082
Loop #1, missing values % after filling with drug column: 10.549619660034407
Loop #1, missing values % after filling with formulary_drug_cd column: 9.868023200005386
Loop #1, missing values % after filling with ndc column: 9.663048752943428

Loop #1, missing values % after re-imploding df: 9.634953412935829
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 9.611727364357368
Loop #2, missing values % after filling with drug column: 9.002189007403166
Loop #2, missing values % after filling with formulary_drug_cd column: 8.987510400175035
Loop #2, missing values % after filling with ndc column: 8.979721405125657

Loop #2, missing values % after re-imploding df: 8.9822875219639
Dataframe exploded, starting for loop #3

Loop #3, missing values

In [75]:
importlib.reload(nachopy)

<module 'nachopy' from 'C:\\Users\\Nacho\\Capstone-project-Brainstation\\nachopy.py'>

In [76]:
nachopy.display_sbs([cat_pres.drug[cat_pres.A<0].value_counts(normalize = True).iloc[0:50]*100],\
                    titles = ['Drugs value counts'],max_rows = 25)


Drugs value counts 
 
 
   
 drug 
 
 
 
 
 isoosmoticdextrose 
 20.622194 
 
 
 bag 
 15.210765 
 
 
 vial 
 7.228978 
 
 
 polyethyleneglycol 
 7.047468 
 
 
 influenzavaccinequadrivalent 
 5.838108 
 
 
 neutraphos 
 5.117210 
 
 
 influenzavirusvaccine 
 5.024741 
 
 
 simethicone 
 3.390417 
 
 
 soln 
 3.070489 
 
 
 pneumococcalvacpolyvalent 
 2.874534 
 
 
 pneumococcalvalentpolysaccharidevaccine 
 2.245615 
 
 
 potassiumphosphate 
 2.128335 
 
 
 phosphorus 
 1.687700 
 
 
 syringe 
 1.650810 
 
 
 sarnalotion 
 1.326720 
 
 
 epoetinalfa 
 1.275467 
 
 
 syringesodiumchloride 
 1.220785 
 
 
 readicatbariumsulfatesuspension 
 1.204054 
 
 
 tiotropiumbromide 
 1.061393 
 
 
 sodiumpolystyrenesulfonate 
 0.832710 
 
 
 prismasatebk 
 0.691680 
 
 
 yellowcaddcassette 
 0.648751 
 
 
 vasopressin 
 0.626879 
 
 
 ferricgluconate 
 0.473526 
 
 
 syringeisoosmoticdextrose 
 0.429291 
 
 
 

 
 
 
   
 drug 
 
 
 
 
 tuberculinprotein 
 0.379098 
 
 
 leucovorincalcium 
 0.338699 
 
 
 gastroviewdiatrizoatemegluminesodium 
 0.327110 
 
 
 lithiumcarbonate 
 0.260594 
 
 
 citratedextroseacdacrrt 
 0.244761 
 
 
 syringechemo 
 0.238722 
 
 
 psylliumwafer 
 0.237089 
 
 
 gammagardliquid 
 0.236436 
 
 
 epiduralbag 
 0.191304 
 
 
 caphosol 
 0.187468 
 
 
 vitaminbcomplex 
 0.186080 
 
 
 hydrocerin 
 0.184693 
 
 
 fosaprepitant 
 0.182489 
 
 
 psyllium 
 0.167472 
 
 
 levalbuterolneb 
 0.148946 
 
 
 aquaphorointment 
 0.145845 
 
 
 tbsyringe 
 0.136459 
 
 
 amp 
 0.134745 
 
 
 psylliumpowder 
 0.129359 
 
 
 mcgvial 
 0.118667 
 
 
 syringesterilewater 
 0.111485 
 
 
 bottle 
 0.099733 
 
 
 protaminesulfate 
 0.096386 
 
 
 fatemulsion 
 0.087001 
 
 
 bluecaddcassette 
 0.085124

Lots of vaccines, let us fill those first

In [77]:
cat = list('ABCDGHJLMNPRSV')
vaccines_bool = (cat_pres.drug.str.contains('vaccine|polyvalent|measlesmumpsrubellavac',regex = True, na = False))
vaccines_bool *= cat_pres.A<0
cat_pres.loc[vaccines_bool,cat] = 0
cat_pres.loc[vaccines_bool,'J'] = 1

print('% of missing values:',nachopy.cat_check(cat_pres)*100)

% of missing values: 7.856632008480606


Let us have a look at the missing values and the drugs per drug type (from the prescriptions table)

In [78]:
print('% of missing values for MAIN drugs',nachopy.cat_check(cat_pres[cat_pres.drug_type =='MAIN'])*100)
print('% of missing values for BASE drugs',nachopy.cat_check(cat_pres[cat_pres.drug_type =='BASE'])*100)
print('% of missing values for ADDITIVE drugs',nachopy.cat_check(cat_pres[cat_pres.drug_type =='ADDITIVE'])*100)

base_drugs_bool = cat_pres.drug_type == 'BASE'
print('\n',cat_pres.drug[base_drugs_bool].drop_duplicates().values)

% of missing values for MAIN drugs 3.185890591466149
% of missing values for BASE drugs 25.785027717319707
% of missing values for ADDITIVE drugs 4.945122703169318

 ['sodiumchloride' 'isoosmoticdextrose' 'sw' 'dextrose' 'bag' 'vial'
 'isoosmoticsodiumchloride' 'ns' 'isotonicsodiumchloride'
 'syringesodiumchloride' 'soln' 'dw' 'syringe' 'sterilewater'
 'syringechemo' 'mcgvial' 'yellowcaddcassette' 'syringeisoosmoticdextrose'
 'epiduralbag' 'amp' 'bluecaddcassette' 'nssyringeintrapleural'
 'syringesterilewater' 'syringelr' 'syringesw' 'nsepiduralbag'
 'clickchangetochooseabasesolution' 'sodiumchlorideexcelbag' 'nssyringe'
 'prismasatebk' 'sodiumchlyellowcadd' 'syringesns' 'bottle' 'syringens'
 'solution' 'sodiumchloridepfsyringe' 'aerogensyringens' 'syringepca'
 'syringeintraperitoneal' 'syringepd' 'dwglassbottle' 'aerogensyringesw'
 'dwwalbumin' 'tbsyringe' 'viaflexbag' 'sodiumcitrate'
 'syringeintraventricular' 'caddcassette' 'syringeinhalation'
 'syringeintrathecal' 'nsintrathecalbag

Rows with drug_type "BASE" correspond with procedures (bag, vial...), solvents (sodium chloride, dextrose) and treatments/drugs (lipid/fat emulsions, travasol, amino acids...) apparently using IV therapy. Many of these procedures will not be very specific per hospital department (sodium chloride, bags, dextrose...), but frequency of procedures could be interesting so registering these events in our table could be useful if we use frequency encoding instead of one-hot encoding. Given that most of these events involve IV therapy and solvents are classed typically both as B (Blood products) and as V (Various) in the ATC system, these prescription events have been classified as such. 

In [79]:
base_drugs_bool = base_drugs_bool * cat_pres.A<0 
cat_pres.loc[base_drugs_bool,cat] = 0

cat_pres.loc[base_drugs_bool,'V'] = 1
cat_pres.loc[base_drugs_bool,'B'] = 1
cat_pres.A.value_counts(normalize = True)*100

print('% of missing values:',nachopy.cat_check(cat_pres)*100)

% of missing values: 4.237109596227292


We still have 4.03% of missing values in our atc1 columns. Let us run the recursive filler function again.


In [80]:
nachopy.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 4.237109596227292
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 4.230159656297725
Loop #1, missing values % after filling with drug column: 4.1881714861804165
Loop #1, missing values % after filling with formulary_drug_cd column: 4.1881714861804165
Loop #1, missing values % after filling with ndc column: 4.18801310429125

Loop #1, missing values % after re-imploding df: 4.179573482857267
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 4.18801310429125
Loop #2, missing values % after filling with drug column: 4.18801310429125
Loop #2, missing values % after filling with formulary_drug_cd column: 4.18801310429125
Loop #2, missing values % after filling with ndc column: 4.18801310429125

Loop #2, missing values % after re-imploding df: 4.179573482857267


We can try and fill manually the drugs that occur most frequently.We can do another round of recursive filling

https://go.drugbank.com/drugs/DB09512

In [81]:
missing_drugs = cat_pres.drug[cat_pres.A<0].value_counts(normalize = True).head(30)
display(missing_drugs*100)
print(f'We will fill {round(missing_drugs.sum()*100,2)}% of the remaining missing drugs manually')
missing_drugsl = missing_drugs.index.values

polyethyleneglycol                      22.796416
neutraphos                              16.552620
simethicone                             10.966969
potassiumphosphate                       6.884517
phosphorus                               5.459197
sarnalotion                              4.291537
epoetinalfa                              4.125747
readicatbariumsulfatesuspension          3.894750
tiotropiumbromide                        3.433283
sodiumpolystyrenesulfonate               2.693563
vasopressin                              2.027762
ferricgluconate                          1.531711
tuberculinprotein                        1.226267
leucovorincalcium                        1.095588
gastroviewdiatrizoatemegluminesodium     1.058100
lithiumcarbonate                         0.842943
psylliumwafer                            0.766912
gammagardliquid                          0.764800
caphosol                                 0.606401
vitaminbcomplex                          0.601913


We will fill 95.78% of the remaining missing drugs manually


In [82]:
cat_pres.loc[cat_pres.drug.isin(missing_drugsl),cat] = 0

cat_pres.loc[cat_pres.drug == 'polyethyleneglycol','B'] = 1
cat_pres.loc[cat_pres.drug == 'neutraphos',['A','B']] = 1
cat_pres.loc[cat_pres.drug == 'simethicone','A'] = 1
cat_pres.loc[cat_pres.drug == 'potassiumphosphate','B'] = 1
cat_pres.loc[cat_pres.drug == 'phosphorus','B'] = 1
cat_pres.loc[cat_pres.drug == 'epoetinalfa','B'] = 1
cat_pres.loc[cat_pres.drug == 'sarnalotion', 'D'] = 1
cat_pres.loc[cat_pres.drug == 'readicatbariumsulfatesuspension',['A','V']] = 1 # Contrast agent
cat_pres.loc[cat_pres.drug == 'tiotropiumbromide','R'] = 1
cat_pres.loc[cat_pres.drug == 'sodiumpolystyrenesulfonate','V'] = 1
cat_pres.loc[cat_pres.drug == 'vasopressin','H'] = 1
cat_pres.loc[cat_pres.drug == 'fishoilmega','C'] = 1
cat_pres.loc[cat_pres.drug == 'ferricgluconate','B'] = 1
cat_pres.loc[cat_pres.drug == 'leucovorincalcium','V'] = 1
cat_pres.loc[cat_pres.drug == 'tuberculinprotein','V'] = 1
cat_pres.loc[cat_pres.drug == 'gastroviewdiatrizoatemegluminesodium',['A','V']] = 1 # Contrast agent
cat_pres.loc[cat_pres.drug == 'lithiumcarbonate','N'] = 1
cat_pres.loc[cat_pres.drug == 'gammagardliquid', 'J'] = 1
cat_pres.loc[cat_pres.drug == 'psylliumwafer', 'A'] = 1
cat_pres.loc[cat_pres.drug == 'caphosol', 'V'] = 1
cat_pres.loc[cat_pres.drug == 'vitaminbcomplex', 'B'] = 1
cat_pres.loc[cat_pres.drug == 'hydrocerin','D'] = 1
cat_pres.loc[cat_pres.drug == 'fosaprepitant','A'] = 1
cat_pres.loc[cat_pres.drug == 'psyllium', 'A'] = 1
cat_pres.loc[cat_pres.drug == 'levalbuterolneb','R'] = 1
cat_pres.loc[cat_pres.drug == 'aquaphorointment','D'] = 1
cat_pres.loc[cat_pres.drug == 'psylliumpowder','A'] = 1
cat_pres.loc[cat_pres.drug == 'protaminesulfate','V'] = 1
cat_pres.loc[cat_pres.drug == 'transplantselfmedicationprogram','V'] = 1
cat_pres.loc[cat_pres.drug == 'bismuthsubsalicylate','A'] = 1
cat_pres.loc[cat_pres.drug == 'xopenexneb','R'] = 1

In [83]:
nachopy.rec_cat_filler(cat_pres,['gsn','drug','formulary_drug_cd','ndc'])

% of missing values before loop: 2.1215980915005153
gsn column split, starting loop...
Dataframe exploded, starting for loop #1

Loop #1, missing values % after filling with gsn column: 2.112882279436674
Loop #1, missing values % after filling with drug column: 2.112650363098966
Loop #1, missing values % after filling with formulary_drug_cd column: 2.111875423141257
Loop #1, missing values % after filling with ndc column: 2.103503808999587

Loop #1, missing values % after re-imploding df: 2.089236509446292
Dataframe exploded, starting for loop #2

Loop #2, missing values % after filling with gsn column: 2.1017955471950027
Loop #2, missing values % after filling with drug column: 2.1017842342029196
Loop #2, missing values % after filling with formulary_drug_cd column: 2.1017842342029196
Loop #2, missing values % after filling with ndc column: 2.1017842342029196

Loop #2, missing values % after re-imploding df: 2.087733298360513
Dataframe exploded, starting for loop #3

Loop #3, missing 

In [84]:
catnas = cat_pres[cat_pres.A < 0]
catnas[['ndc','drug_type','drug','formulary_drug_cd','gsn']].isna().sum()/len(catnas)*100

ndc                   7.399508
drug_type            94.689407
drug                 96.884561
formulary_drug_cd    96.884561
gsn                  95.162184
dtype: float64

In [85]:
cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).iloc[50]*100

0.3180704508443245

In [86]:
nachopy.display_sbs([cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).iloc[:60]*100],\
                    titles = ['ndc value counts'], max_rows = 10)
            

missing_ndcs = cat_pres.ndc[cat_pres.A<0].value_counts(normalize = True).head(30)
print(f'We could fill {round(missing_ndcs.sum()*100,2)}% of the remaining missing ndcs manually by filling the first 30 ndcs')

ndc value counts 
 
 
   
 ndc 
 
 
 
 
 100030116 
 7.928287 
 
 
 105850865 
 7.021317 
 
 
 100030115 
 6.088819 
 
 
 105720810 
 4.999633 
 
 
 107700128 
 4.999340 
 
 
 353560215 
 2.331245 
 
 
 109390331 
 1.748507 
 
 
 109390803 
 1.675151 
 
 
 123330025 
 1.600035 
 
 
 109390021 
 1.284019 
 
 
 

 
 
 
   
 ndc 
 
 
 
 
 328490500 
 1.182788 
 
 
 100390049 
 1.134080 
 
 
 103100000 
 1.087719 
 
 
 109390170 
 0.970936 
 
 
 118222280 
 0.822758 
 
 
 328490706 
 0.749696 
 
 
 118225764 
 0.742067 
 
 
 103270035 
 0.715072 
 
 
 109160046 
 0.704508 
 
 
 101350203 
 0.661669 
 
 
 

 
 
 
   
 ndc 
 
 
 
 
 109390232 
 0.656387 
 
 
 614420501 
 0.639369 
 
 
 109390026 
 0.622350 
 
 
 328490707 
 0.616482 
 
 
 122800008 
 0.596235 
 
 
 109390197 
 0.551048 
 
 
 461220143 
 0.502633 
 
 
 107540669 
 0.495885 
 
 
 101350120 
 0.477399 
 
 
 109390378 
 0.450111 
 
 
 

 
 
 
   
 ndc 
 
 
 
 
 109390789 
 0.445123 
 
 
 118220148 
 0.444242 
 
 
 116940846 
 0.429278 
 
 
 109390556 
 0.415780 
 
 
 101190021 
 0.401696 
 
 
 107120478 
 0.400229 
 
 
 106310098 
 0.399055 
 
 
 118450060 
 0.394654 
 
 
 108760691 
 0.392013 
 
 
 118224607 
 0.389079 
 
 
 

 
 
 
   
 ndc 
 
 
 
 
 101350683 
 0.388492 
 
 
 109390008 
 0.378809 
 
 
 370000143 
 0.373527 
 
 
 328490708 
 0.368539 
 
 
 122800077 
 0.366192 
 
 
 110262618 
 0.361790 
 
 
 543383 
 0.348586 
 
 
 499990990 
 0.336263 
 
 
 102670737 
 0.333622 
 
 
 103700511 
 0.322178 
 
 
 

 
 
 
   
 ndc 
 
 
 
 
 116940301 
 0.318070 
 
 
 4071412 
 0.312202 
 
 
 491000400 
 0.311615 
 
 
 118229932 
 0.298998 
 
 
 165900587 
 0.298705 
 
 
 109390454 
 0.293423 
 
 
 240900492 
 0.288435 
 
 
 118450127 
 0.270829 
 
 
 118220146 
 0.270829 
 
 
 154820234 
 0.268189

We could fill 54.06% of the remaining missing ndcs manually by filling the first 30 ndcs


The problem is that the missing values are more distributed between many ndcs, so filling manually 30 of them would just tackle 53.83% of the missing categories by ndc. Moreover, browsing ndc is not so easy, as they are specific per supplier and can be harder to find. Even more if they could not be found on the FDA website. We will leave these values unfilled for now. We can save the merged, filled prescriptions df to be used later.

In [88]:
cat.append('join_on_index')
cat_pres = cat_pres[cat].set_index('join_on_index',drop = True)
prescriptions = prescriptions[['subject_id','hadm_id','starttime','join_on_index']].set_index('join_on_index', drop = True)
prescriptions_final = prescriptions.join(cat_pres)

In [90]:
prescriptions_final.to_pickle('../Capstone data/pickles/prescriptions_final.pkl')

In [ ]:
non_null_cats = cat_pres.A>=0
cat_pres = cat_pres.loc[non_null_cats,cat]
prescriptions = prescriptions[['subject_id','hadm_id','starttime']]
prescriptions = prescriptions.join(cat_pres)
prescriptions.reset_index(drop = True, inplace = True)

In [ ]:
cat_pres.A.value_counts(normalize = True)*100

In [ ]:
unclass_drugs = cat_pres.drug[cat_pres.A<0].drop_duplicates()
udrugs_gsnvc = cat_pres.gsn[cat_pres.A<0].value_counts()
gsn_udrugs = udrugs_gsnvc[udrugs_gsnvc>1].index.values

In [ ]:
cat.append('gsn')
cat_gsn = cat_pres.loc[cat_pres.gsn.isin(gsn_udrugs),cat].groupby('gsn').max()
cat_gsn = cat_gsn[cat_gsn.A>-1]
cat_gsn

In [ ]:
cat = cat[:-1]
cat_pres[cat] = cat_pres[cat].where(cat_pres[cat]>=0)

In [ ]:
cat_pres.set_index(ind,append = True)

In [ ]:
cat_pres.set_index('gsn', inplace = True)
cat_pres.update(cat_gsn)
cat_pres.set_index(ind)

In [ ]:
cat_pres.head()

In [ ]:
sum(cat_pres.drug.str.contains('polyvalent'))

In [ ]:
cat_pres.gsn[misspelt_drugs_bool].value_counts().describe()

In [ ]:
(5303415+7791711)/len(cat_pres)

In [ ]:
cat

In [ ]:
cat = list('ABCDGHJLMNPRSV')
cat.append('drug')
cat_drug = cat_pres[cat].fillna(0.0001).groupby('formulary_drug_cd',dropna = False).sum()
cat_drug

In [ ]:
atr4_pres.atc4_name.nunique()

The missing values labelled as 0 are now empty strings due to the conversion. There may be other values that are not 0, maybe even some could be in other ndc formats (hyphenated or with less digits). We can find out how many rows have which number of digits.

So all fields that do not have 9-digit numbers have no digits. We can consider all of these as missing values.

More than 12% of ndc fields are missing. Some of the drugs listed here may not have an ndc (Sodium Chloride flush on 2nd row for instance), but others may be mistakes. NDC is specific to product, not to drug (same drug from different companies will have a different code) so filling from the table would not be very accurate. Given that what we need the NDC just to classify our drugs, we will leave these fields blank and will fill the categories instead after NDC-matching.


In [ ]:
prescriptions.ndc[~missing_ndc].drop_duplicates().to_csv('Unique_NDCs.csv')

Let us see if we can fill these missing values by using the other fields in the table. pharmacy_id, gsn, drug and formulary_drug_cd could be used for this.

In [ ]:
prescriptions.formulary_drug_cd.info()

In [ ]:
fdrugna = prescriptions[prescriptions.formulary_drug_cd.isna()]

In [ ]:
def df_filler(x):
    

prescriptions.formulary_drug_cd[prescriptions.drug.isin(fdrugna.drug)].isna()

In [ ]:
prescriptions.gsn.value_counts()

In [ ]:
prescriptions.pharmacy_id.value_counts()

In [ ]:
prescriptions.formulary_drug_cd.value_counts()

In [ ]:
prescriptions.formulary_drug_cd.info()

In [ ]:
prescriptions.drug = prescriptions.drug.str.lower()
prescriptions.drug.value_counts().describe()

In [ ]:
ndc_bool = prescriptions.ndc!=u'' & prescriptions 
print(sum(ndc_bool))
print(sum(~ndc_bool))

In [ ]:
elswhr_bool = prescriptions.loc[~ndc_bool,'drug'].isin(prescriptions.loc[ndc_bool,'drug'])
print(round(sum(elswhr_bool)/sum(~ndc_bool)*100,2))

In [ ]:
prescriptions.loc[elswhr_bool,'ndc'] = prescriptions.loc[ndc_bool*prescriptions.,'ndc']

In [ ]:
prescriptions.ndc.nunique()

In [ ]:
import numpy as np

In [ ]:
medication = pd.read_json(r'C:\Users\Nacho\Downloads\Medication.ndjson', lines = True)

In [ ]:
medication.shape

In [ ]:
prescriptions.ndc.head()

In [ ]:
ndc_desc = pd.read_csv("G:\My Drive\Brainstation\Capstone project\product.csv",usecols = ['PRODUCTNDC','NONPROPRIETARYNAME','PHARM_CLASSES'])

In [ ]:
ndc_desc = ndc_desc.loc[~ndc_desc.PHARM_CLASSES.isna()].reset_index(drop = True)

In [ ]:
ndc_desc.head()

In [ ]:
def sht2nineNDC(x):
    x = x.split('-')
    x = x[0].zfill(5) + x[1].zfill(4)
    
    return x
    

ndc_desc.PRODUCTNDC = ndc_desc.PRODUCTNDC.apply(sht2nineNDC)

In [ ]:
ndc_desc.PRODUCTNDC.head()

In [ ]:
classpername = ndc_desc[['NONPROPRIETARYNAME','PHARM_CLASSES']].groupby('NONPROPRIETARYNAME').nunique()

In [ ]:
classperndc = ndc_desc[['PRODUCTNDC','PHARM_CLASSES']].groupby('PRODUCTNDC').nunique()

In [ ]:
classpername.value_counts()

In [ ]:
classperndc.value_counts()

In [ ]:
np.mean(prescriptions.ndc.isna())*100

In [ ]:
prescriptions.loc[prescriptions.ndc.isna()].head(10)

In [ ]:
ndc_desc.rename(columns = {'PRODUCTNDC' : 'ndc'},inplace = True)

In [ ]:
presfinal = pd.merge(prescriptions,ndc_desc, how = 'left',on = 'ndc')

In [ ]:
np.mean(presfinal.PHARM_CLASSES.isna())

In [ ]:
np.mean(presfinal.NONPROPRIETARYNAME.isna())

In [ ]:
presfinal[['ndc','drug','NONPROPRIETARYNAME','PHARM_CLASSES']]

In [ ]:
presfinal.hadm_id.nunique()

In [ ]:
len(presfinal.PHARM_CLASSES.isna())-sum(presfinal.PHARM_CLASSES.isna())

In [ ]:
classperhadm = presfinal[['PHARM_CLASSES','hadm_id']].groupby('hadm_id').nunique()
classperhadm.describe()

In [ ]:
classperhadm.loc[classperhadm.PHARM_CLASSES == 0].count()

In [ ]:
prescriptions.ndc.apply(lambda x: str(x)[:-2].zfill(9))

In [ ]:
print(ndc_desc.NONPROPRIETARYNAME.nunique())
print(ndc_desc_clean.NONPROPRIETARYNAME.nunique())

In [ ]:
ndc_desc.shape

In [ ]:
ndc_desc.PRODUCTNDC.apply(lambda x: int(x.replace('-','')))

In [ ]:
ndc_desc.PHARM_CLASSES.value_counts

In [ ]:
poe = pd.read_csv('MIMIC IV ED/poe.csv')

In [ ]:
poe.order_status.value_counts(normalize = True)

In [ ]:
poe.loc[poe.order_type == 'Lab'].order_subtype.value_counts()

In [ ]:
poe.order_type.value_counts()

In [ ]:
admissions = pd.read_csv('MIMIC IV ED/admissions.csv')

In [ ]:
admissions.race.value_counts(normalize = True)

In [ ]:
sum(admissions.admission_location.isna())

In [ ]:
admissions.admission_location.value_counts(normalize = True)

In [ ]:
transfers = pd.read_csv('MIMIC IV ED/transfers.csv')

In [ ]:
sum(transfers.eventtype == 'discharge')

In [ ]:
transfers.careunit.value_counts(normalize = True)

In [ ]:
transfers.careunit.unique()

In [ ]:
transfers[['intime','outtime']] = pd.to_datetime(transfers[['intime', 'outtime']])

In [ ]:
transfers.info(show_counts = True)

In [ ]:
bool_outs

In [ ]:
transfers.describe()

In [ ]:
bool_discharge = transfers.eventtype=='discharge'
bool_carenan = transfers.careunit.isna()
bool_outtime = transfers.outtime.isna()

In [ ]:
print(all(bool_discharge == bool_carenan))
print(all(bool_discharge == bool_outtime))


In [ ]:
transfers.head()

In [ ]:
services = pd.read_csv('MIMIC IV ED/services.csv')

In [ ]:
services.shape

In [ ]:
services.head()

In [ ]:
services.prev_service.value_counts()

In [ ]:
transfers.eventtype.value_counts()

In [ ]:
transfers.

In [ ]:
emar = pd.read_csv('MIMIC IV/emar.csv')

In [ ]:
emar.medication.nunique()

In [ ]:
emar.head()

In [ ]:
transfers.careunit.value_counts()

In [ ]:
print('Number of patients:',transfers['subject_id'].nunique())
print('Number of transfers (and rows number):',transfers['transfer_id'].nunique())
print('Number of stays:',transfers['hadm_id'].nunique())
print('Number of departments:',transfers['careunit'].nunique())

In [ ]:
sum(transfers['careunit'].isna())

In [ ]:
print('Number of departments:',transfers['careunit'].unique())

In [ ]:
admissions = pd.read_csv('MIMIC IV ED/admissions.csv')

In [ ]:
pd.to_datetime(admissions.admittime,infer_datetime_format = True).describe(datetime_is_numeric = True)

In [ ]:
admissions.discharge_location.unique()

In [ ]:
admissions.head()

In [ ]:
len(admissions)

In [ ]:
print('Number of patients:',admissions['subject_id'].nunique())
print('Number of transfers (and length of rows):',admissions['transfer_id'].nunique())
print('Number of stays:',transfers['hadm_id'].nunique())
print('Number of departments:',transfers['careunit'].nunique())

In [ ]:
edstays['disposition'].value_counts()

In [ ]:
diagnosis.icd_version.value_counts()

In [ ]:
diagnosis['icd_title'].value_counts().head()

In [ ]:
print('\033[1m' + 'diagnosis')
display(diagnosis.head())
display(diagnosis.info())
print('\033[1m' + 'edstays')
display(edstays.head())
display(diagnosis.info())
print('\033[1m' + 'medrecon')
display(medrecon.head())
display(medrecon.info())
print('\033[1m' + 'pyxis')
display(pyxis.head())
display(pyxis.info())
print('\033[1m' + 'triage')
display(triage.head())
display(triage.info())
print('\033[1m' + 'vitalsign')
display(vitalsign.head())
display(vitalsign.info())